File: leetcode/22-generate-parentheses.py
# https://leetcode.com/problems/generate-parentheses/ (top-interview-questions)

from typing import List

class Solution:
	def generateParenthesis(self, n: int) -> List[str]:
		
		parentheses = list()
		
		def backtracking(stack = list(), opened = 0, closed = 0):
			if len(stack) == 2*n: # opened == closed == n
				parentheses.append(''.join(stack))
				return
			if opened < n:
				stack.append('(')
				backtracking(stack, opened+1, closed)
				stack.pop()
			if closed < opened:
				stack.append(')')
				backtracking(stack, opened, closed+1)
				stack.pop()

		backtracking()
		return parentheses

File: leetcode/22-generate-parentheses.js
// https://leetcode.com/problems/generate-parentheses/ (top-interview-questions)

/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
  const combinations = [];
  const parentheses = [];

  const backtracking = (opened, closed) => {
    if ( opened === closed && opened === n ) {
      combinations.push(parentheses.join(''));
      return;
    }

    if (opened < n) {
      parentheses.push('(');
      backtracking(opened + 1, closed);
      parentheses.pop();
    }
    if (closed < opened) {
      parentheses.push(')');
      backtracking(opened, closed+1);
      parentheses.pop();
    }
  }

  backtracking(0, 0);
  
  return combinations;
};

const result = generateParenthesis(8);
console.log(result);